protocol/daemons/server/types/health_checker.go

```diff
@@ -0,0 +1,169 @@
+func (u *healthCheckerMutableState) ReportSuccess(now time.Time) {
+	u.lock.Lock()
+	defer u.lock.Unlock()
+
+	u.mostRecentSuccess = now
+
+	// Whenever the service is healthy, reset the first failure in streak timestamp.
+	u.mostRecentFailureStreakError.Reset()
+}
+
+// ReportFailure updates the health checker's mutable state to reflect a failed health check and schedules the next
+// poll as an atomic operation. The method returns the duration of the current failure streak.
+func (u *healthCheckerMutableState) ReportFailure(now time.Time, err error) time.Duration {
+	u.lock.Lock()
+	defer u.lock.Unlock()
+
+	u.mostRecentFailureStreakError.Update(now, err)
+
+	return now.Sub(u.mostRecentFailureStreakError.Timestamp())
+}
+
+// healthChecker encapsulates the logic for monitoring the health of a health checkable service.
+type healthChecker struct {
+	// mutableState is the mutable state of the health checker. Access to these fields is synchronized.
+	mutableState *healthCheckerMutableState
+
+	// healthCheckable is the health checkable service to be monitored.
+	healthCheckable types.HealthCheckable
+
+	// pollFrequency is the frequency at which the health checkable service is polled.
+	pollFrequency time.Duration
+
+	// timer triggers a health check poll for a health checkable service.
+	timer *time.Timer
+
+	// maxAcceptableUnhealthyDuration is the maximum acceptable duration for a health checkable service to
+	// remain unhealthy. If the service remains unhealthy for this duration, the monitor will execute the
+	// specified callback function.
+	maxAcceptableUnhealthyDuration time.Duration
+
+	// unhealthyCallback is the callback function to be executed if the health checkable service remains
+	// unhealthy for a period of time greater than or equal to the maximum acceptable unhealthy duration.
+	// This callback function is executed with the error that caused the service to become unhealthy.
+	unhealthyCallback func(error)
+
+	// timeProvider is used to get the current time. It is added as a field so that it can be mocked in tests.
+	timeProvider libtime.TimeProvider
+
+	logger log.Logger
+}
+
+// Poll executes a health check for the health checkable service. If the service has been unhealthy for longer than the
+// maximum acceptable unhealthy duration, the callback function is executed.
+// This method is publicly exposed for testing. This method is synchronized.
+func (hc *healthChecker) Poll() {
+	err := hc.healthCheckable.HealthCheck()
+	now := hc.timeProvider.Now()
+
+	// Capture healthy response.
+
+	if err == nil {
+		hc.mutableState.ReportSuccess(now)
+	} else {
+		streakDuration := hc.mutableState.ReportFailure(now, err)
+		// If the service has been unhealthy for longer than the maximum acceptable unhealthy duration, execute the
+		// callback function.
+		if streakDuration >= hc.maxAcceptableUnhealthyDuration {
+			hc.unhealthyCallback(err)
+		}
+	}
+
+	// Schedule next poll.
+	hc.timer.Reset(hc.pollFrequency)
```

Comment:
Nice, thanks for simplifying this logic!

```suggestion
	if err == nil { // Capture healthy response.
		hc.mutableState.ReportSuccess(now)
	} else { // Capture error response.
		failureDuration := hc.mutableState.ReportFailure(now, err)
		// If the service has been unhealthy for longer than the maximum acceptable unhealthy duration, execute the
		// callback function.
		if failureDuration >= hc.maxAcceptableUnhealthyDuration {
			hc.unhealthyCallback(err)
		}
	}

	// Schedule next poll.
	hc.timer.Reset(hc.pollFrequency)
```